File: tests/typesafety/python/api_parse.py
# expect: passes
# title: pydantic and zod check the order at the door
# improves: api_dict
from pydantic import BaseModel


class Order(BaseModel):
    id: str
    quantity: int
    gift_note: str | None = None


def price_cents(order: Order) -> int:
    return order.quantity * 250  # no checks: the type says both


def handle(body: str) -> int:
    order = Order.model_validate_json(body)
    return price_cents(order)

File: tests/typesafety/typescript/api_parse.ts
// expect: passes

import { z } from "zod";

const Order = z.object({
  id: z.string(),
  quantity: z.number().int(),
  giftNote: z.string().nullable(),
});

type Order = z.infer<typeof Order>;

function priceCents(order: Order): number {
  return order.quantity * 250; // no checks: the type says both
}

export function handle(body: string): number {
  const order = Order.parse(JSON.parse(body));
  return priceCents(order);
}
